From 3ce89e0315c42f3faf3530d130552ffcd30a2d54 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon at codeweavers.com>
Date: Wed, 17 Nov 2021 11:51:47 +0100
Subject: [PATCH] win32u: Initialize unused_entry on font allocation.

There may be a race condition otherwise between release_gdi_font and
find_cached_gdi_font, leading to invalid memory access:

One thread calling release_gdi_font may decrement refcount to 0,
then try to enter font_lock. At the same time, another thread may be
calling find_cached_gdi_font through select_font, holding the font_lock.

This second thread would find refcount set to 0, and then try to remove
unused_entry from its list, although it hasn't been added yet to the
unused list.

Another possible way to fix this would be to hold the font_lock before
decrementing refcount, but that may not be as efficient.

Signed-off-by: R?mi Bernon <rbernon at codeweavers.com>
---
 dlls/win32u/font.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/win32u/font.c b/dlls/win32u/font.c
index d2f61526540..437ce7862a8 100644
--- a/dlls/win32u/font.c
+++ b/dlls/win32u/font.c
@@ -1989,6 +1989,7 @@ static struct gdi_font *alloc_gdi_font( const WCHAR *file, void *data_ptr, SIZE_
     font->matrix.eM11 = font->matrix.eM22 = 1.0;
     font->scale_y = 1;
     font->kern_count = -1;
+    list_init( &font->unused_entry );
     list_init( &font->child_fonts );
 
     if (file)
-- 
2.33.1

